cassandra/src/main/scala/geotrellis/spark/io/cassandra/CassandraIndexing.scala

```diff
@@ -0,0 +1,272 @@
+    keyspace: String,
+    table: String,
+    layerName: String,
+    zoom: Int
+  ): WriteValueStatement = {
+    indexStrategy match {
+      case conf.ReadOptimized =>
+        WriteValueStatement(
+          QueryBuilder
+            .insertInto(keyspace, table)
+            .value("name", layerName)
+            .value("zoom", zoom)
+            .value("zoombin", QueryBuilder.bindMarker())
+            .value("key", QueryBuilder.bindMarker())
+            .value("value", QueryBuilder.bindMarker())
+        )
+
+      case conf.WriteOptimized =>
+        WriteValueStatement(
+          QueryBuilder
+            .insertInto(keyspace, table)
+            .value("name", layerName)
+            .value("zoom", zoom)
+            .value("key", QueryBuilder.bindMarker())
+            .value("value", QueryBuilder.bindMarker())
+        )
+    }
+  }
+
+  private def zoomBin(
+    index: BigInteger
+  ): java.lang.Integer = {
+    zoomBinIntervals.intervals.find{ case (interval, idx) => interval.contains(index) }.map {
+      _._2: java.lang.Integer
+    }.getOrElse(0: java.lang.Integer)
+  }
+
+  def prepareQuery(
+    statement: CassandraStatement
+  )(session: => Session): CassandraPreparedStatement = {
+    statement match {
+      case WriteValueStatement(s) => PreparedWriteValueStatement(session.prepare(s))
+      case QueryValueStatement(s) => PreparedQueryValueStatement(session.prepare(s))
+      case DeleteStatement(s) => PreparedDeleteStatement(session.prepare(s))
+    }
+  }
+
+  def bindQuery(
+    queryStrategy: CassandraIndexStrategy,
+    statement: CassandraPreparedStatement,
+    index: BigInteger,
+    value: Option[ByteBuffer] = None
+  ): BoundStatement = {
+    queryStrategy match {
+      case conf.ReadOptimized =>
+        statement match{
+          case PreparedQueryValueStatement(p) => p.bind(zoomBin(index), index)
+          case PreparedDeleteStatement(p) => p.bind(zoomBin(index), index)
+          case PreparedWriteValueStatement(p) =>
+            //TODO: Is there a better way to enforce this than at runtime?
+            require(value.nonEmpty, "Cannot bind write statement with empty value")
+            p.bind(zoomBin(index), index, value.get)
+        }
+      case conf.WriteOptimized =>
+        statement match {
+          case PreparedQueryValueStatement(p) => p.bind(index)
+          case PreparedDeleteStatement(p) => p.bind(index)
+          case PreparedWriteValueStatement(p) =>
+            require(value.nonEmpty, "Cannot bind write statement with empty value")
+            p.bind(index, value.get)
+        }
+    }
+  }
+
+}
+
+object CassandraIndexing{
+  sealed trait CassandraStatement extends Any
+  sealed trait CassandraPreparedStatement extends Any
+
+  //Value Classes - should be unwrapped at compile time:
+  case class ZoomBinIntervals(intervals: Vector[(Interval[BigInt], Int)]) extends AnyVal
+
+  final case class WriteValueStatement(statement: BuiltStatement) extends AnyVal with CassandraStatement
```

Comment:
Trying to add in some type safety here... Would have reached for tagged types, but I wasn't sure how the project felt about them (or which library to use for them).  I opted for the native-supported "value classes" construct instead... which relies upon the compiler to do the "right thing" a little too much for my taste, but should still avoid some of the (un)boxing if I understand correctly.